common/utils/remoteServiceConnector.go

```diff
@@ -0,0 +1,109 @@
+package utils
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+// RemoteServiceConnector is a convenient struct to help implementing GetClientDataDeserializer().
+// Here is an example of usage when you want to serialize a Connection and a string.
+//
+//	 type ClientData struct {
+//	     MyString   string
+//	     Connection utils.Connection
+//	 }
+//
+//	 type MyRemoteService {
+//	     // This implement GetClientDataDeserializer() func(auto.UpResult) (*ClientData, error)
+//	     *RemoteServiceConnector[ClientData]
+//	 }
+//
+//		remoteServiceConnector := utils.NewRemoteServiceConnector(ctx, ClientData{})
+//		remoteServiceConnector.Register(stackKey, "Connection", connection)
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities. -->
**💭 thought**
Would be cool if `Register` was accepting a `pulumi.Input` version of `ClientData`, with a matching type matrix
```
string <-> pulumi.StringInput
int <-> pulumi.IntInput
utils.Connection <-> remote.Connection
```

Reply:
I don't know how to do it :). Also this code will be used by a very limited number of users and a very limited number of times, so I think this is fine.